emily/handler/src/api/handlers/internal.rs

```diff
@@ -0,0 +1,220 @@
+}
+
+/// Sets the api status to the provided status.
+///
+/// Return meanings:
+/// - Err(e):
+///     Something went wrong.
+/// - Ok(None):
+///     The API status is already what we wanted it to be, so there's
+///     no action required.
+/// - Ok(Some(ApiStateEntry)):
+///     We have successfully converted the api to the state returned.
+async fn set_api_state_status(
+    context: &EmilyContext,
+    new_status: &ApiStatus,
+) -> Result<Option<ApiStateEntry>, Error> {
+    let mut update_attempts = 0;
+    let mut api_state: ApiStateEntry;
+    loop {
+        update_attempts += 1;
+        let original_api_state = accessors::get_api_state(context).await?;
+        api_state = original_api_state.clone();
+
+        // Update the api status.
+        api_state.api_status = match (new_status, &original_api_state.api_status) {
+            // Handle trying to set the api status to reorganizing.
+            (ApiStatus::Reorg(_), ApiStatus::Stable(_)) => new_status.clone(),
+            (ApiStatus::Reorg(new_reorg_tip), ApiStatus::Reorg(current_reorg_tip)) => {
+                if new_reorg_tip == current_reorg_tip {
+                    return Ok(None);
+                } else {
+                    warn!(
+                        "Trying to reorg with new chaintip {:?} while the api is reorganizing around the chaintip {:?}",
+                        new_reorg_tip,
+                        current_reorg_tip,
+                    );
+                    return Err(Error::InconsistentState(Inconsistency::ItemUpdate(
+                        format!("Trying to reorg with new chaintip {:?} while the api is reorganizing around the chaintip {:?}",
+                        new_reorg_tip,
+                        current_reorg_tip,
+                        ))));
+                }
+            }
+            (ApiStatus::Stable(new_tip), ApiStatus::Stable(old_tip)) => {
+                if new_tip == old_tip {
+                    return Ok(None);
+                } else {
+                    new_status.clone()
+                }
+            }
+            (ApiStatus::Stable(_), ApiStatus::Reorg(_)) => new_status.clone(),
+        };
+
+        debug!(
+            "Changing Api state from [{:?}] to [{:?}]. Attempt {} of maximum {}.",
+            original_api_state, api_state, update_attempts, MAX_SET_API_STATE_ATTEMPTS_DURING_REORG,
+        );
+
+        // Attempt to set the API state.
+        match accessors::set_api_state(context, &api_state).await {
+            // We successfully set the API state.
+            Ok(()) => {
+                info!("Successfully set api state: {:?}.", api_state);
+                break;
+            }
+            // Retry if there was a version conflict.
+            Err(Error::VersionConflict) => {
+                if update_attempts >= MAX_SET_API_STATE_ATTEMPTS_DURING_REORG {
+                    debug!("Failed to update API state {:?}", api_state);
+                    return Err(Error::InternalServer);
```

Comment:
I prefer internal server error because more has gone wrong than a version conflict internally here than just a failed update, even if that was the reason for the deeper internal failure.